mmpy_bot/plugins/base.py

```diff
@@ -88,21 +70,213 @@ async def call_function(
+                )
+            else:
+                if not h.function_docheader:
+                    string += f"- `{cmd}` {direct} {mention}\n"
+                else:
+                    string += f"- `{cmd}` {direct} {mention} - {h.function_docheader}\n"
 
         return string
 
+    def get_help(self, message: Message):
+        """Obtain Help info from PluginManager.
+
+        Override this method if you need to customize which listeners will be included
+        in the help.
+        """
+        return self.plugin_manager.get_help()
+
+    @listen_to("^help$", needs_mention=True)
     async def help(self, message: Message):
-        """Prints the list of functions registered on every active plugin."""
-        self.driver.reply_to(message, self.get_help_string())
+        """Shows this help information."""
+        self.driver.reply_to(
+            message, self.get_help_string(message), direct=self.direct_help
+        )
+
+
+@dataclass
+class FunctionInfo:
+    help_type: str
+    location: str
+    function: Function
+    pattern: str
+    plugin_docheader: str
+    plugin_docfull: str
+    function_docheader: str
+    function_docfull: str
+    direct: bool
+    mention: bool
+    is_click: bool
+    metadata: Dict
+
+
+def generate_plugin_help(
+    listeners: Dict[re.Pattern[Any], List[Union[MessageFunction, WebHookFunction]]],
+):
+    """Build FunctionInfo objects from plugin and function information.
+
+    Returns one FunctionInfo instance for every listener (message or webhook)
+    """
+
+    plug_help: List[FunctionInfo] = []
+
+    for matcher, functions in listeners.items():
+        for function in functions:
+            plug_head, plug_full = split_docstring(function.plugin.__doc__)
+            func_head, func_full = split_docstring(function.docstring)
+
+            if isinstance(function, MessageFunction):
+                direct = function.direct_only
+                mention = function.needs_mention
+                help_type = "message"
+            elif isinstance(function, WebHookFunction):
+                direct = mention = False
+                help_type = "webhook"
+            else:
+                raise NotImplementedError(
+                    f"Unknown/Unsupported listener type: '{type(function)}'"
+                )
+
+            plug_help.append(
+                FunctionInfo(
+                    help_type=help_type,
+                    location=function.plugin.__class__.__name__,
+                    function=function,
+                    pattern=matcher.pattern,
+                    plugin_docheader=plug_head,
+                    plugin_docfull=plug_full,
+                    function_docheader=func_head,
+                    function_docfull=func_full,
+                    direct=direct,
+                    mention=mention,
+                    is_click=function.is_click_function,
+                    metadata=function.metadata,
```

Comment:
Since all of these attributes are accessible on the `Function` itself, I don't see the added value of this `FunctionInfo` class. Instead of storing this all in that class here, you might as well change the `HelpPlugin` to directly obtain this from the `Function`?

Reply:
Okay, fair enough :+1: